compiler/rustc_arena/src/lib.rs

```diff
@@ -110,6 +110,28 @@ impl<T> Default for TypedArena<T> {
 }
 
 impl<T> TypedArena<T> {
+    #[inline]
+    unsafe fn write_from_iter<I: Iterator<Item = T>>(
+        &self,
```

Comment:
self is unused here. Maybe make this function module local instead of duplicating it?

Reply:
I actually tried that and ended up with an issue where the lifetime of the returned slice couldn't be inferred, and I gave up on it quickly. Right now, let's wait for the perf results first if this helps at all :)

Reply:
Yeah my original suggestion was that the function gets used by both Typed and Dropless arenas.

Reply:
Hmmm in order to fix the unsoundness as pointed out below it might be needed to use `self` in `write_from_iter` after all, so I guess it can't be deduplicated after all...

Reply:
Not sure, since `write_from_iter` passes back a slice of the exact size, that size can be used to move the pointer.